EIPS/eip-3540.md

```diff
@@ -0,0 +1,218 @@
+
+The section contents follow after the header, in the order and size they are defined, without any padding bytes.
+
+To summarise, the bytecode has the following format:
+```
+format, magic, version, (section_kind, section_size)+, 0, <section contents>
+```
+
+#### Validation rules
+
+A bytestream starting with the *EOF prefix* declares itself conforming to the rules according to its version.
+
+1. The rules of `version=1` are specified below:
+- `section_size` MUST NOT be 0.
+- Exactly one code section MUST be present.
+- The code section MUST be the first section.
+- A single data section MAY follow the code section.
+- Stray bytes outside of sections MUST NOT be present. This includes trailing bytes after the last section.
+2. Any other version is invalid.
+
+(*Note:* Contract creation code SHOULD set the section size of the data section so that the constructor arguments fit it.)
+
+#### Changes to execution semantics
+
+For clarity, the *container* refers to the complete account code, while *code* refers to the contents of the code section only.
+
+1. Jumpdest analysis is only run on the *code*.
+2. Execution starts at the first byte of the *code*, and `PC` is set to this position within the container format (e.g. `PC=10` for a *container* with a code and data section).
+3. If `PC` goes outside of the code section bounds, execution  aborts with failure.
+4. `PC` returns the current position within the *container*.
+5. `JUMP`/`JUMPI` uses an absolute offset within the *container*.
+6. `CODECOPY`/`CODESIZE`/`EXTCODECOPY`/`EXTCODESIZE`/`EXTCODEHASH` keeps operating on the entire *container*.
+7. The input to `CREATE`/`CREATE2` is still the entire *container*.
+
+#### Changes to contract creation semantics
+
+For clarity, the *EOF prefix* together with a version number $v$ is denoted as the *EOF$v$ prefix*, e.g. *EOF1 prefix*. 
+
+1. If _initcode's container_ has EOF1 prefix it must be valid EOF1 code.
+2. If _code's container_ has EOF1 prefix it must be valid EOF1 code.
+
+## Rationale
+
+EVM and/or account versioning has been discussed numerous times over the past years. This proposal aims to learn from them. A good starting point collecting the previous proposals is [this thread](https://ethereum-magicians.org/t/ethereum-account-versioning/3508).
+
+### Execution vs. creation time validation
+
+This specification introduces creation time validation, which means:
+- All created contracts with EOF$n$ prefix are valid according to version $n$ rules. This is very strong and useful property. The client can trust that the deployed code is well-formed.
+- In future, this allows serializing `JUMPDEST` map in the EOF container and eliminate the need of implicit `JUMPDEST` analysis required before execution.
+- Or completely remove the need for `JUMPDEST` instructions.
+- This helps with deprecating EVM instructions and/or features.
+- The biggest disadvantage is that deploy-time validation of EOF code must be enabled in two hard-forks.
```

Comment:
Remove the multi-hard fork mentions, as they are out of scope for this specification and in scope for hard fork coordination discussion.
```suggestion
```

Reply:
Hard fork coordination complexities can affect the design desicions of the spec, can't they?

Reply:
I suppose since this is the rationale section this line is probably fine to leave.  I still standby that the HF coordination stuff should be removed from the specification, and it should be limited in the rationale.  The key thing to remember is that EIPs are technical specifications designed to be read by future developers and they shouldn't overly focus on the discussions of *today*, but instead around how to make the specification understandable to developers of the future.